lib/ansible/modules/crypto/openssh_keypair.py

```diff
@@ -0,0 +1,317 @@
+filename:
+    description: Path to the generated SSH private key file
+    returned: changed or success
+    type: string
+    sample: /tmp/id_ssh_rsa
+fingerprint:
+    description: The fingerprint of the key.
+    returned: changed or success
+    type: string
+    sample: 4096 SHA256:r4YCZxihVjedH2OlfjVGI6Y5xAYtdCwk8VxKyzVyYfM example@example.com (RSA)
+'''
+
+import os
+import errno
+import subprocess
+
+from subprocess import PIPE, Popen
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+
+
+class KeypairError(Exception):
+    pass
+
+
+class Keypair(object):
+
+    def __init__(self, module):
+        self.path = module.params['path']
+        self.state = module.params['state']
+        self.force = module.params['force']
+        self.size = module.params['size']
+        self.type = module.params['type']
+        self.comment = module.params['comment']
+        self.changed = False
+        self.check_mode = module.check_mode
+        self.privatekey = None
+        self.fingerprint = {}
+
+        if self.type in ('rsa', 'rsa1'):
+            self.size = 4096 if self.size is None else self.size
+            if self.size < 1024:
+                module.fail_json(msg=('For C(RSA) keys, the minimum size is 1024 bits and the default is 4096 bits. '
+                                      'Attempting to use bit lengths under 1024 will cause the module to fail.'))
+
+        if self.type == 'dsa':
+            self.size = 1024 if self.size is None else self.size
+            if self.size != 1024:
+                module.fail_json(msg=('C(DSA) keys must be exactly 1024 bits as specified by FIPS 186-2.'))
+
+        if self.type == 'ecdsa':
+            self.size = 256 if self.size is None else self.size
+            if self.size not in (256, 384, 521):
+                module.fail_json(msg=('For C(ECDSA) keys, size determines the key length by selecting from '
+                                      'one of three elliptic curve sizes: 256, 384 or 521 bits. '
+                                      'Attempting to use bit lengths other than these three values for '
+                                      'C(ECDSA) keys will cause this module to fail. '))
+        if self.type == 'ed25519':
+            self.size = 256
+
+    def generate(self, module):
+        # generate a keypair
+        if not self.check(module, perms_required=False) or self.force:
+            args = [
+                'ssh-keygen',
+                '-q',
+                '-N', '',
+                '-b', str(self.size),
+                '-t', self.type,
+                '-f', self.path,
+            ]
+
+            if self.comment:
+                args.extend(['-C', self.comment])
+            else:
+                args.extend(['-C', ""])
+
+            try:
+                self.changed = True
+                subprocess.call(args)
+                proc = Popen(['ssh-keygen', '-lf', self.path], stdout=PIPE)
```

Comment:
Out of curiosity: why aren't you using `module.run_command()` instead of directly using `subprocess.call` and `Popen`?